src/System.Buffers.ReaderWriter/System/Buffers/Reader/BufferReader_text.cs

```diff
@@ -58,42 +64,88 @@ private unsafe static bool TryParseSlow(ref BufferReader<byte> reader, out bool
         {
-            const int MaxLength = 15;
-
-            ReadOnlySpan<byte> unread = reader.UnreadSpan;
+            // Note that ints can have leading zeros ("-00000000000000127") which makes predicting the max
+            // length difficult. Typically the int will be less than 15 characters ("-2,147,483,648") so we'll
+            // work in those increments.
+            int bufferSize = 15;
 
-            if (unread.Length > MaxLength)
+            // Try using the stack for scratch space.
+            do
             {
-                // Fast path had enough space but couldn't find valid data
-                value = default;
-                return false;
-            }
-
-            byte* buffer = stackalloc byte[MaxLength];
-            Span<byte> tempSpan = new Span<byte>(buffer, MaxLength);
-            if (Utf8Parser.TryParse(reader.PeekSlow(tempSpan), out value, out int consumed))
+                // stackalloc is separated as we can't represent that PeekSlow will not capture the span.
+                byte* stackBuffer = stackalloc byte[bufferSize];
+                Span<byte> buffer = new Span<byte>(stackBuffer, bufferSize);
+                ReadOnlySpan<byte> peekSpan = reader.PeekSlow(buffer);
+
+                if (!Utf8Parser.TryParse(peekSpan, out value, out int consumed, standardFormat))
+                {
+                    // Overflow
+                    return false;
+                }
+
+                if (consumed < peekSpan.Length || peekSpan.Length < buffer.Length)
+                {
+                    // The parser found a digit it wouldn't consume or we hit the end of the reader
+                    // (peekSpan was smaller than requested)so we have all valid data.
+                    reader.Advance(consumed);
+                    return true;
+                }
+
+
+                bufferSize *= 2;
+            } while (bufferSize <= 60);
+
+            // Start heap allocating to get enough space.
+            do
             {
-                reader.Advance(consumed);
-                return true;
-            }
-
-            return false;
+                byte[] buffer = new byte[bufferSize];
+                ReadOnlySpan<byte> peekSpan = reader.PeekSlow(buffer);
+
+                if (!Utf8Parser.TryParse(peekSpan, out value, out int consumed, standardFormat))
+                {
+                    // Overflow
+                    return false;
+                }
+
+                if (consumed < peekSpan.Length || peekSpan.Length < buffer.Length)
+                {
+                    // The parser found a digit it wouldn't consume or we hit the end of the reader
+                    // (peekSpan was smaller than requested)so we have all valid data.
+                    reader.Advance(consumed);
+                    return true;
+                }
+
+                // Throw if doubling would overflow.
+                bufferSize = checked(bufferSize * 2);
+            } while (true);
```

Comment:
should we break out the loop after some threshold?

Reply:
We'll throw eventually with an overflow above. I'm not sure we can pick anything more rational than that. We could, in theory, return false- but we're starting to hit crazytown levels here (MBs of leading zeroes). I'm open to suggestions here.

Reply:
What does int.Parse do? Does it accept 1 billion leading zeros?

Reply:
Well, up to almost int.MaxValue in theory. I considered doing our own scanning to get past these cases, but it is complicated and would fall over if the Utf8Parser changes behavior or adds another format type. There are a number of corner cases in that code.

I think it is reasonable to expect users to do their own skipping for crazy cases. We've got all the helpers for that.